### PR TITLE
feat: use aws sdk3 EI-722

### PR DESCRIPTION
**Pull request by James1218**
## Upgrade aws sdk from v2 to v3
This PR does the following:
1. Upgraded aws sdk from v2 to v3.
2. Fixed the broken `scan` function.
3. Fixed broken tests. 
4. Added an optional `config` argument in the `Table` constructor to allow setting regions other than the default region. e.g. `const table = new Table("test", { region: "us-west-2"})`

## Testing Done
The unit tests passed.
```
 PASS  test/__test__/query.test.ts (7.658s)
 PASS  test/__test__/delete.test.ts (7.731s)
 PASS  test/__test__/put.test.ts (7.788s)
 PASS  test/__test__/get.test.ts (7.778s)
 PASS  test/__test__/update.test.ts (7.918s)
 PASS  test/__test__/scan.test.ts (8.142s)

Test Suites: 6 passed, 6 total
Tests:       37 passed, 37 total
```

## The missing "between" function
The `between` function will be done in a separate PR.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,21 +23,22 @@
...
-    "big.js": "^6.2.1",
-    "chai": "^4.3.7",
-    "jest": "^29.3.1",
-    "prettier": "^2.8.1",
-    "serverless": "^3.26.0",
-    "serverless-dynamodb-local": "^0.2.40",
-    "ts-jest": "^29.0.3",
-    "tslint": "^5.20.1",
+    "@types/jest": "^24.0.17",
+    "@types/node": "^12.6.9",
```
Comment (liushuyu) [inline package.json:31]:
```suggestion
    "@types/node": "^18.11.17",
```

Do we want to downgrade Node.js typing to `^12`? Node.js 12 is already EOL, and we should target our Node.js typing data to `16.x` or `18.x`.

File: package.json [hunk 2]
```diff
@@ -23,21 +23,22 @@
...
+    "chai": "^4.2.0",
+    "jest": "^24.8.0",
+    "prettier": "^1.18.2",
+    "serverless": "^1.49.0",
+    "serverless-dynamodb-local": "^0.2.38",
+    "ts-jest": "^24.0.2",
+    "tslint": "^5.18.0",
     "tslint-config-prettier": "^1.18.0",
-    "typescript": "^4.9.4"
+    "typescript": "^3.9.10"
```
Comment (liushuyu) [inline package.json:41]:
```suggestion
    "typescript": "^4.9.4"
```

I am not quite sure why TypeScript compile is downgraded here.

File: src/CRUD/delete.ts [hunk 3]
```diff
@@ -15,7 +15,7 @@ export default async function deleteItem({
   key,
   options = {},
   verbose = false,
-  forTrx = false,
+  forTrx = false
```
Comment (liushuyu) [inline src/CRUD/delete.ts:18]: I do think we probably need to introduce a Prettier code-style rule file in this repository, especially given this is an open-source repository.

File: src/index.ts [hunk 4]
```diff
@@ -20,36 +29,49 @@ interface IIndex {
...
+    } else {
+      this.dbClient = dbClient;
+      this.docClient = docClient;
+    }
   }
 
-  public async initTable() {
-    const tableInfo: any = await dynamodb.describeTable({ TableName: this.tableName }).promise();
+  public async initTable(): Promise<void> {
+    // const tableInfo: any = await dynamodb.describeTable({ TableName: this.tableName }).promise();
```
Comment (liushuyu) [inline src/index.ts:64]:
```suggestion
```

Usually, we want to clean up any code-comments since we have the version control system (Git).

**Review by liushuyu — APPROVED** [review]
Overall should be good to merge as-is, but I have some concerns.

File: package.json (hunk 1 above)
Comment (James1218) [reply package.json]: Upgraded back to v18. I reverted package.json to the working commit, so a bulk a packages were downgraded.

File: package.json (hunk 2 above)
Comment (James1218) [reply package.json]: Upgraded back to v4. I reverted package.json to the working commit, so a bulk a packages were downgraded.

File: src/CRUD/delete.ts (hunk 3 above)
Comment (James1218) [reply src/CRUD/delete.ts:18]: There is a prettier config file in the root. The dot was removed after I ran `npm run format`. Feel free to add more rules to the existing `.prettierrc`

File: src/index.ts (hunk 4 above)
Comment (James1218) [reply src/index.ts]: Cleanup is done.

File: test/__test__/query.test.ts [hunk 5]
```diff
@@ -1,52 +1,57 @@
...
+      secretAccessKey: "test",
+    },
   };
-  const localDynamodb = new AWS.DynamoDB(dynamoOptions);
-  const localClient = new AWS.DynamoDB.DocumentClient(dynamoOptions);
+  const localDbClient = new DynamoDBClient(dynamoOptions);
+  const localDocClient = DynamoDBDocumentClient.from(localDbClient);
 
-  Table.replaceDynamoClient(localDynamodb, localClient);
+  Table.replaceDynamoClient(localDbClient, localDocClient);
```
Comment (ChenLi0830) [inline test/__test__/query.test.ts:21]: Thanks for the MR! Could you add a couple of tests to test the new optional config param in the constructor?

Namely, something like this
```
tableWithSmallData = new Table("Clevo-Processed-Speech-Table", {region: "ap-northeast-1"});
``` 
And test the region is indeed changed?

**Review by ChenLi0830 — APPROVED** [review]
Looks good! Approved with a comment about testing the new `config`

File: test/__test__/query.test.ts (hunk 5 above)
Comment (James1218) [reply test/__test__/query.test.ts:23]: New tests were added. 
For each operation, 2 new tests were added to test default region and override region, respectively.

Comment (James1218) [reply test/__test__/query.test.ts:23]: e.g. 
1. test default region: https://github.com/Eon-Labs/dynamo-light/pull/36/files#diff-90b0fdc25ec525fa2f62184130c1a7e750018144d97abf7171d6aa71ddb20b9bR75
2. test override region: https://github.com/Eon-Labs/dynamo-light/pull/36/files#diff-90b0fdc25ec525fa2f62184130c1a7e750018144d97abf7171d6aa71ddb20b9bR89

**Pull request merged**